tests/pipeline/test_quarters_estimates.py

```diff
@@ -0,0 +1,898 @@
+        quarter-event where estimate1 < estimate2 and all estimates are < the
+        relevant event and assign each of these inter-leavings to a
+        different sid.
+        """
+
+        sid_estimates = []
+        sid_releases = []
+        # We want all permutations of 2 knowledge dates per quarter.
+        it = enumerate(
+            itertools.permutations(cls.q1_knowledge_dates +
+                                   cls.q2_knowledge_dates,
+                                   4)
+        )
+        for sid, (q1e1, q1e2, q2e1, q2e2) in it:
+            # We're assuming that estimates must come before the relevant
+            # release.
+            if (q1e1 < q1e2 and
+                    q2e1 < q2e2 and
+                    # All estimates are < Q2's event, so just constrain Q1
+                    # estimates.
+                    q1e1 < cls.q1_release_dates[0] and
+                    q1e2 < cls.q1_release_dates[0]):
+                sid_estimates.append(cls.create_estimates_df(q1e1,
+                                                             q1e2,
+                                                             q2e1,
+                                                             q2e2,
+                                                             sid))
+                sid_releases.append(cls.create_releases_df(sid))
+
+        return pd.concat(sid_estimates + sid_releases).reset_index(drop=True)
+
+    @classmethod
+    def create_releases_df(cls, sid):
+        # Final release dates never change. The quarters have very tight date
+        # ranges in order to reduce the number of dates we need to iterate
+        # through when testing.
+        return pd.DataFrame({
+            TS_FIELD_NAME: [pd.Timestamp('2015-01-15'),
+                            pd.Timestamp('2015-01-31')],
+            EVENT_DATE_FIELD_NAME: [pd.Timestamp('2015-01-15'),
+                                    pd.Timestamp('2015-01-31')],
+            'estimate': [0.5, 0.8],
+            FISCAL_QUARTER_FIELD_NAME: [1.0, 2.0],
+            FISCAL_YEAR_FIELD_NAME: [2015.0, 2015.0],
+            SID_FIELD_NAME: sid
+        })
+
+    @classmethod
+    def create_estimates_df(cls,
+                            q1e1,
+                            q1e2,
+                            q2e1,
+                            q2e2,
+                            sid):
+        return pd.DataFrame({
+            EVENT_DATE_FIELD_NAME: cls.q1_release_dates + cls.q2_release_dates,
+            'estimate': [.1, .2, .3, .4],
+            FISCAL_QUARTER_FIELD_NAME: [1.0, 1.0, 2.0, 2.0],
+            FISCAL_YEAR_FIELD_NAME: [2015.0, 2015.0, 2015.0, 2015.0],
+            TS_FIELD_NAME: [q1e1, q1e2, q2e1, q2e2],
+            SID_FIELD_NAME: sid,
+        })
+
+    @classmethod
+    def init_class_fixtures(cls):
+        # Must be generated before call to super since super uses `events`.
+        cls.events = cls.gen_estimates()
```

Comment:
Why does super use events? Shouldn't super just cache this? Could we make it a classlazyval maybe? In general you should only add attributes after the call to super.